### PR TITLE
tsdb: Save allocations on labels by re-using label array

### PR DESCRIPTION
**Pull request by brancz**
I believe this should fix https://github.com/prometheus/prometheus/issues/7732.

@bwplotka @codesome @brian-brazil

### CODE REVIEW

**Comment by roidelapluie** [comment]
Is this prombenchworthy?

**Comment by bwplotka** [comment]
No, but it would be nice to run Select bench on it. It looks good to me (:

**Comment by roidelapluie** [comment]
I have the funcbench malediction...

**Comment by krasi-georgiev** [comment]
there is an update for funcbench that should solve this. will be merged today

**Comment by brancz** [comment]
Are we actually calling this correctly?

**Comment by roidelapluie** [comment]
Yes but it needs an update that will arrive today.

**Comment by brancz** [comment]
Anything I can do to help it along? :)

**Comment by bwplotka** [comment]
Yes @brancz it would be appreciated if someone has time to run `BenchmarkQuerierSelect` locally and compare with and without this PR :hugs: 

That's the only thing left (: There is a rule: **Never trust your eyes in terms of perf improvements until you see results** ;p

**Comment by brancz** [comment]
@bwplotka I meant the funcbench "fix" :)

**Comment by brancz** [comment]
Looks mixed:

```
benchmark                                                old ns/op      new ns/op      delta
BenchmarkQuerierSelect/Head/1of1000000-8                 466893848      380383208      -18.53%
BenchmarkQuerierSelect/Head/10of1000000-8                497267937      412377178      -17.07%
BenchmarkQuerierSelect/Head/100of1000000-8               407210963      394276352      -3.18%
BenchmarkQuerierSelect/Head/1000of1000000-8              392726174      386735189      -1.53%
BenchmarkQuerierSelect/Head/10000of1000000-8             396210322      420197867      +6.05%
BenchmarkQuerierSelect/Head/100000of1000000-8            428185499      470494794      +9.88%
BenchmarkQuerierSelect/Head/1000000of1000000-8           671177681      738573544      +10.04%
BenchmarkQuerierSelect/SortedHead/1of1000000-8           1775712879     2086305777     +17.49%
BenchmarkQuerierSelect/SortedHead/10of1000000-8          1735133140     1855856583     +6.96%
BenchmarkQuerierSelect/SortedHead/100of1000000-8         1798964578     1928457124     +7.20%
BenchmarkQuerierSelect/SortedHead/1000of1000000-8        1687464406     2082599465     +23.42%
BenchmarkQuerierSelect/SortedHead/10000of1000000-8       1681772608     2354247158     +39.99%
BenchmarkQuerierSelect/SortedHead/100000of1000000-8      1698929506     2027262925     +19.33%
BenchmarkQuerierSelect/SortedHead/1000000of1000000-8     1959577713     2138919743     +9.15%
BenchmarkQuerierSelect/Block/1of1000000-8                574979728      567111772      -1.37%
BenchmarkQuerierSelect/Block/10of1000000-8               583562770      577580506      -1.03%
BenchmarkQuerierSelect/Block/100of1000000-8              600302107      567982871      -5.38%
BenchmarkQuerierSelect/Block/1000of1000000-8             575752036      580213844      +0.77%
BenchmarkQuerierSelect/Block/10000of1000000-8            622611077      810807972      +30.23%
BenchmarkQuerierSelect/Block/100000of1000000-8           698878044      1016014379     +45.38%
BenchmarkQuerierSelect/Block/1000000of1000000-8          1128960288     1140317753     +1.01%
```

**Comment by krasi-georgiev** [comment]
> Are we actually calling funcbench correctly?

yes the issue is somewhere else, when you provide wrong args to funcbench it will post back a comment to notify about the wrong usage and also prints a link to the readme for the correct usage

**Comment by brancz** [comment]
Tried a few more things, latest benchmark is now:

```
○ benchcmp old.txt new.txt                                                                                             
benchcmp is deprecated in favor of benchstat: https://pkg.go.dev/golang.org/x/perf/cmd/benchstat                       
benchmark                                                old ns/op      new ns/op      delta    
BenchmarkQuerierSelect/Head/1of1000000-8                 296799961      285314344      -3.87%                          
BenchmarkQuerierSelect/Head/10of1000000-8                287250120      287423672      +0.06%   
BenchmarkQuerierSelect/Head/100of1000000-8               289651377      283307649      -2.19%                          
BenchmarkQuerierSelect/Head/1000of1000000-8              287464786      290702025      +1.13%   
BenchmarkQuerierSelect/Head/10000of1000000-8             289590086      288229907      -0.47%
BenchmarkQuerierSelect/Head/100000of1000000-8            307543377      302647535      -1.59%
BenchmarkQuerierSelect/Head/1000000of1000000-8           462260730      430338135      -6.91%
BenchmarkQuerierSelect/SortedHead/1of1000000-8           1101255421     1078921062     -2.03%
BenchmarkQuerierSelect/SortedHead/10of1000000-8          1103327856     1095498903     -0.71%
BenchmarkQuerierSelect/SortedHead/100of1000000-8         1088456364     1087725571     -0.07%
BenchmarkQuerierSelect/SortedHead/1000of1000000-8        1091687651     1098109147     +0.59%
BenchmarkQuerierSelect/SortedHead/10000of1000000-8       1092085332     1089343240     -0.25%
BenchmarkQuerierSelect/SortedHead/100000of1000000-8      1116357490     1113042790     -0.30%
BenchmarkQuerierSelect/SortedHead/1000000of1000000-8     1255759984     1215396251     -3.21%
BenchmarkQuerierSelect/Block/1of1000000-8                501893927      489486589      -2.47%
BenchmarkQuerierSelect/Block/10of1000000-8               500186117      486818372      -2.67%
BenchmarkQuerierSelect/Block/100of1000000-8              508082676      499404936      -1.71%
BenchmarkQuerierSelect/Block/1000of1000000-8             502062640      500199044      -0.37%
BenchmarkQuerierSelect/Block/10000of1000000-8            512914331      500545053      -2.41%
BenchmarkQuerierSelect/Block/100000of1000000-8           536542749      513072876      -4.37% 
BenchmarkQuerierSelect/Block/1000000of1000000-8          719338478      616495805      -14.30%
```

tl;dr huge queries get faster, small queries stay roughly the same

**Comment by geekodour** [comment]
/funcbench master BenchmarkQuerierSelect

hoping this works as expected! :pray:

**Comment by brancz** [comment]
I don't understand the funcbench output, when literally running the same command gives me this:

```
benchmark                                                old ns/op      new ns/op      delta
BenchmarkQuerierSelect/Head/1of1000000-8                 292395617      280572216      -4.04%
BenchmarkQuerierSelect/Head/10of1000000-8                288081239      288336309      +0.09%
BenchmarkQuerierSelect/Head/100of1000000-8               293925265      281808322      -4.12%
BenchmarkQuerierSelect/Head/1000of1000000-8              294774636      290407317      -1.48%
BenchmarkQuerierSelect/Head/10000of1000000-8             292633875      286288476      -2.17%
BenchmarkQuerierSelect/Head/100000of1000000-8            306918348      301298413      -1.83%
BenchmarkQuerierSelect/Head/1000000of1000000-8           463544980      423725674      -8.59%
BenchmarkQuerierSelect/SortedHead/1of1000000-8           1126705413     1099482013     -2.42%
BenchmarkQuerierSelect/SortedHead/10of1000000-8          1126151936     1113746536     -1.10%
BenchmarkQuerierSelect/SortedHead/100of1000000-8         1117822136     1108380530     -0.84%
BenchmarkQuerierSelect/SortedHead/1000of1000000-8        1131196263     1103403956     -2.46%
BenchmarkQuerierSelect/SortedHead/10000of1000000-8       1117592481     1112906157     -0.42%
BenchmarkQuerierSelect/SortedHead/100000of1000000-8      1136235830     1111508887     -2.18%
BenchmarkQuerierSelect/SortedHead/1000000of1000000-8     1280433034     1228725261     -4.04%
BenchmarkQuerierSelect/Block/1of1000000-8                513837151      531370128      +3.41%
BenchmarkQuerierSelect/Block/10of1000000-8               514763484      525483317      +2.08%
BenchmarkQuerierSelect/Block/100of1000000-8              518187861      529311642      +2.15%
BenchmarkQuerierSelect/Block/1000of1000000-8             521457922      526518763      +0.97%
BenchmarkQuerierSelect/Block/10000of1000000-8            529431477      513692094      -2.97%
BenchmarkQuerierSelect/Block/100000of1000000-8           545169522      532103644      -2.40%
BenchmarkQuerierSelect/Block/1000000of1000000-8          745794602      664299538      -10.93%

benchmark                                                old allocs     new allocs     delta
BenchmarkQuerierSelect/Head/1of1000000-8                 1000015        1000013        -0.00%
BenchmarkQuerierSelect/Head/10of1000000-8                1000051        1000040        -0.00%
BenchmarkQuerierSelect/Head/100of1000000-8               1000411        1000310        -0.01%
BenchmarkQuerierSelect/Head/1000of1000000-8              1004011        1003010        -0.10%
BenchmarkQuerierSelect/Head/10000of1000000-8             1040011        1030010        -0.96%
BenchmarkQuerierSelect/Head/100000of1000000-8            1400011        1300010        -7.14%
BenchmarkQuerierSelect/Head/1000000of1000000-8           5000010        4000010        -20.00%
BenchmarkQuerierSelect/SortedHead/1of1000000-8           1000052        1000050        -0.00%
BenchmarkQuerierSelect/SortedHead/10of1000000-8          1000087        1000077        -0.00%
BenchmarkQuerierSelect/SortedHead/100of1000000-8         1000447        1000347        -0.01%
BenchmarkQuerierSelect/SortedHead/1000of1000000-8        1004047        1003047        -0.10%
BenchmarkQuerierSelect/SortedHead/10000of1000000-8       1040047        1030047        -0.96%
BenchmarkQuerierSelect/SortedHead/100000of1000000-8      1400047        1300047        -7.14%
BenchmarkQuerierSelect/SortedHead/1000000of1000000-8     5000047        4000047        -20.00%
BenchmarkQuerierSelect/Block/1of1000000-8                3000016        3000012        -0.00%
BenchmarkQuerierSelect/Block/10of1000000-8               3000050        3000030        -0.00%
BenchmarkQuerierSelect/Block/100of1000000-8              3000410        3000210        -0.01%
BenchmarkQuerierSelect/Block/1000of1000000-8             3004010        3002010        -0.07%
BenchmarkQuerierSelect/Block/10000of1000000-8            3040010        3020010        -0.66%
BenchmarkQuerierSelect/Block/100000of1000000-8           3400010        3200010        -5.88%
BenchmarkQuerierSelect/Block/1000000of1000000-8          7000010        5000010        -28.57%

benchmark                                                old bytes     new bytes     delta
BenchmarkQuerierSelect/Head/1of1000000-8                 48000784      48000992      +0.00%
BenchmarkQuerierSelect/Head/10of1000000-8                48002368      48002288      -0.00%
BenchmarkQuerierSelect/Head/100of1000000-8               48018208      48015248      -0.01%
BenchmarkQuerierSelect/Head/1000of1000000-8              48176608      48144848      -0.07%
BenchmarkQuerierSelect/Head/10000of1000000-8             49760608      49440848      -0.64%
BenchmarkQuerierSelect/Head/100000of1000000-8            65600608      62400848      -4.88%
BenchmarkQuerierSelect/Head/1000000of1000000-8           224000546     192000850     -14.29%
BenchmarkQuerierSelect/SortedHead/1of1000000-8           101191800     101192008     +0.00%
BenchmarkQuerierSelect/SortedHead/10of1000000-8          101193320     101193304     -0.00%
BenchmarkQuerierSelect/SortedHead/100of1000000-8         101209160     101206264     -0.00%
BenchmarkQuerierSelect/SortedHead/1000of1000000-8        101367560     101335864     -0.03%
BenchmarkQuerierSelect/SortedHead/10000of1000000-8       102951560     102631864     -0.31%
BenchmarkQuerierSelect/SortedHead/100000of1000000-8      118791560     115591864     -2.69%
BenchmarkQuerierSelect/SortedHead/1000000of1000000-8     277191560     245191864     -11.54%
BenchmarkQuerierSelect/Block/1of1000000-8                99200840      99200984      +0.00%
BenchmarkQuerierSelect/Block/10of1000000-8               99202472      99202136      -0.00%
BenchmarkQuerierSelect/Block/100of1000000-8              99219752      99213656      -0.01%
BenchmarkQuerierSelect/Block/1000of1000000-8             99392552      99328856      -0.06%
BenchmarkQuerierSelect/Block/10000of1000000-8            101120552     100480856     -0.63%
BenchmarkQuerierSelect/Block/100000of1000000-8           118400552     112000856     -5.41%
BenchmarkQuerierSelect/Block/1000000of1000000-8          291200552     227200856     -21.98%
```

**Comment by geekodour** [comment]
funcbench output uses [benchstat](https://godoc.org/golang.org/x/perf/cmd/benchstat) whereas your output i am assuming is using [benchcmp](https://godoc.org/golang.org/x/tools/cmd/benchcmp) , additionally @bwplotka suggested using `-delta=none` in https://github.com/prometheus/test-infra/pull/400 but current output is not using `-delta=none`

I'll create a post in mailing list to get feedbacks on output format etc today. :)

**Comment by bwplotka** [comment]
`-delta=none` is much more usable for our cases I believe.

**Comment by brancz** [comment]
gotcha, with `-delta-test=none` it now looks like this:

```
name                                         old time/op    new time/op    delta
QuerierSelect/Head/1of1000000-8                 292ms ± 0%     281ms ± 0%   -4.04%
QuerierSelect/Head/10of1000000-8                288ms ± 0%     288ms ± 0%   +0.09%
QuerierSelect/Head/100of1000000-8               294ms ± 0%     282ms ± 0%   -4.12%
QuerierSelect/Head/1000of1000000-8              295ms ± 0%     290ms ± 0%   -1.48%
QuerierSelect/Head/10000of1000000-8             293ms ± 0%     286ms ± 0%   -2.17%
QuerierSelect/Head/100000of1000000-8            307ms ± 0%     301ms ± 0%   -1.83%
QuerierSelect/Head/1000000of1000000-8           464ms ± 0%     424ms ± 0%   -8.59%
QuerierSelect/SortedHead/1of1000000-8           1.13s ± 0%     1.10s ± 0%   -2.42%
QuerierSelect/SortedHead/10of1000000-8          1.13s ± 0%     1.11s ± 0%   -1.10%
QuerierSelect/SortedHead/100of1000000-8         1.12s ± 0%     1.11s ± 0%   -0.84%
QuerierSelect/SortedHead/1000of1000000-8        1.13s ± 0%     1.10s ± 0%   -2.46%
QuerierSelect/SortedHead/10000of1000000-8       1.12s ± 0%     1.11s ± 0%   -0.42%
QuerierSelect/SortedHead/100000of1000000-8      1.14s ± 0%     1.11s ± 0%   -2.18%
QuerierSelect/SortedHead/1000000of1000000-8     1.28s ± 0%     1.23s ± 0%   -4.04%
QuerierSelect/Block/1of1000000-8                514ms ± 0%     531ms ± 0%   +3.41%
QuerierSelect/Block/10of1000000-8               515ms ± 0%     525ms ± 0%   +2.08%
QuerierSelect/Block/100of1000000-8              518ms ± 0%     529ms ± 0%   +2.15%
QuerierSelect/Block/1000of1000000-8             521ms ± 0%     527ms ± 0%   +0.97%
QuerierSelect/Block/10000of1000000-8            529ms ± 0%     514ms ± 0%   -2.97%
QuerierSelect/Block/100000of1000000-8           545ms ± 0%     532ms ± 0%   -2.40%
QuerierSelect/Block/1000000of1000000-8          746ms ± 0%     664ms ± 0%  -10.93%

name                                         old alloc/op   new alloc/op   delta
QuerierSelect/Head/1of1000000-8                48.0MB ± 0%    48.0MB ± 0%   +0.00%
QuerierSelect/Head/10of1000000-8               48.0MB ± 0%    48.0MB ± 0%   -0.00%
QuerierSelect/Head/100of1000000-8              48.0MB ± 0%    48.0MB ± 0%   -0.01%
QuerierSelect/Head/1000of1000000-8             48.2MB ± 0%    48.1MB ± 0%   -0.07%
QuerierSelect/Head/10000of1000000-8            49.8MB ± 0%    49.4MB ± 0%   -0.64%
QuerierSelect/Head/100000of1000000-8           65.6MB ± 0%    62.4MB ± 0%   -4.88%
QuerierSelect/Head/1000000of1000000-8           224MB ± 0%     192MB ± 0%  -14.29%
QuerierSelect/SortedHead/1of1000000-8           101MB ± 0%     101MB ± 0%   +0.00%
QuerierSelect/SortedHead/10of1000000-8          101MB ± 0%     101MB ± 0%   -0.00%
QuerierSelect/SortedHead/100of1000000-8         101MB ± 0%     101MB ± 0%   -0.00%
QuerierSelect/SortedHead/1000of1000000-8        101MB ± 0%     101MB ± 0%   -0.03%
QuerierSelect/SortedHead/10000of1000000-8       103MB ± 0%     103MB ± 0%   -0.31%
QuerierSelect/SortedHead/100000of1000000-8      119MB ± 0%     116MB ± 0%   -2.69%
QuerierSelect/SortedHead/1000000of1000000-8     277MB ± 0%     245MB ± 0%  -11.54%
QuerierSelect/Block/1of1000000-8               99.2MB ± 0%    99.2MB ± 0%   +0.00%
QuerierSelect/Block/10of1000000-8              99.2MB ± 0%    99.2MB ± 0%   -0.00%
QuerierSelect/Block/100of1000000-8             99.2MB ± 0%    99.2MB ± 0%   -0.01%
QuerierSelect/Block/1000of1000000-8            99.4MB ± 0%    99.3MB ± 0%   -0.06%
QuerierSelect/Block/10000of1000000-8            101MB ± 0%     100MB ± 0%   -0.63%
QuerierSelect/Block/100000of1000000-8           118MB ± 0%     112MB ± 0%   -5.41%
QuerierSelect/Block/1000000of1000000-8          291MB ± 0%     227MB ± 0%  -21.98%

name                                         old allocs/op  new allocs/op  delta
QuerierSelect/Head/1of1000000-8                 1.00M ± 0%     1.00M ± 0%   -0.00%
QuerierSelect/Head/10of1000000-8                1.00M ± 0%     1.00M ± 0%   -0.00%
QuerierSelect/Head/100of1000000-8               1.00M ± 0%     1.00M ± 0%   -0.01%
QuerierSelect/Head/1000of1000000-8              1.00M ± 0%     1.00M ± 0%   -0.10%
QuerierSelect/Head/10000of1000000-8             1.04M ± 0%     1.03M ± 0%   -0.96%
QuerierSelect/Head/100000of1000000-8            1.40M ± 0%     1.30M ± 0%   -7.14%
QuerierSelect/Head/1000000of1000000-8           5.00M ± 0%     4.00M ± 0%  -20.00%
QuerierSelect/SortedHead/1of1000000-8           1.00M ± 0%     1.00M ± 0%   -0.00%
QuerierSelect/SortedHead/10of1000000-8          1.00M ± 0%     1.00M ± 0%   -0.00%
QuerierSelect/SortedHead/100of1000000-8         1.00M ± 0%     1.00M ± 0%   -0.01%
QuerierSelect/SortedHead/1000of1000000-8        1.00M ± 0%     1.00M ± 0%   -0.10%
QuerierSelect/SortedHead/10000of1000000-8       1.04M ± 0%     1.03M ± 0%   -0.96%
QuerierSelect/SortedHead/100000of1000000-8      1.40M ± 0%     1.30M ± 0%   -7.14%
QuerierSelect/SortedHead/1000000of1000000-8     5.00M ± 0%     4.00M ± 0%  -20.00%
QuerierSelect/Block/1of1000000-8                3.00M ± 0%     3.00M ± 0%   -0.00%
QuerierSelect/Block/10of1000000-8               3.00M ± 0%     3.00M ± 0%   -0.00%
QuerierSelect/Block/100of1000000-8              3.00M ± 0%     3.00M ± 0%   -0.01%
QuerierSelect/Block/1000of1000000-8             3.00M ± 0%     3.00M ± 0%   -0.07%
QuerierSelect/Block/10000of1000000-8            3.04M ± 0%     3.02M ± 0%   -0.66%
QuerierSelect/Block/100000of1000000-8           3.40M ± 0%     3.20M ± 0%   -5.88%
QuerierSelect/Block/1000000of1000000-8          7.00M ± 0%     5.00M ± 0%  -28.57%
```

**Comment by nevill** [comment]
> `-delta=none` is much more usable for our cases I believe.

The major feature of `benchstat` is to tell if a change is significant. e.g. we don't care the change in Millisecond, even the difference is bigger than 10%. We only care if the change is in Second scale.

**Comment by bwplotka** [comment]
Ah - looks like then someone need to teach us how to read this (:

**Comment by brian-brazil** [comment]
> we don't care the change in Millisecond, even the difference is bigger than 10%

That seems unwise for microbenchmarks, where the relative change is what mainly matters. Whether a change matters in absolute terms is a usually a judgement call that we need to make.

**Comment by nevill** [comment]
OK, means we need to have an option to add `-delta-test=none` in funcbench.

> > we don't care the change in Millisecond, even the difference is bigger than 10%
> 
> That seems unwise for microbenchmarks, where the relative change is what mainly matters. Whether a change matters in absolute terms is a usually a judgement call that we need to make.
